File: leetcode/59.spiral-matrix-ii/cpp/generatematrix_test.cpp
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include "doctest.h"

#include "generatematrix.cpp"

struct testCase
{
    int n;
    std::vector<std::vector<int>> want;
};

TEST_CASE("GenerateMatrix")
{
    std::vector<testCase> tcs{
        {3, {{1, 2, 3}, {8, 9, 4}, {7, 6, 5}}},
        {1, {{1}}},
    };

    Solution s;

    for (auto tc : tcs)
    {
        auto got = s.generateMatrix(tc.n);
        CHECK(got == tc.want);
    }
}

File: leetcode/59.spiral-matrix-ii/go/generatematrix_test.go
package generatematrix

import (
	"fmt"
	"reflect"
	"testing"
)

type testcase struct {
	n    int
	want [][]int
}

func TestGenerateMatrix(t *testing.T) {
	tcs := []testcase{
		{n: 3, want: [][]int{[]int{1, 2, 3}, []int{8, 9, 4}, []int{7, 6, 5}}},
		{n: 1, want: [][]int{[]int{1}}},
	}

	for _, tc := range tcs {
		t.Run(fmt.Sprintf("%d -> %v", tc.n, tc.want), func(t *testing.T) {
			got := generateMatrix(tc.n)

			if !reflect.DeepEqual(got, tc.want) {
				t.Fatalf("generatematrix, n: %d, want: %v, got: %v", tc.n, tc.want, got)
			}
		})

	}
}
